### PR TITLE
Refactor node querying logic

### PR DESCRIPTION
**Pull request by prettyirrelevant**
## Checklist
This PR does the following:

- Upon node connection, its capabalities are determined i.e whether it's pruned, archive node or not.
- Refactor typing of node methods e.g `get_transaction_by_hash`, `get_transaction_receipt`
- Handle cases where a tx hash does not exist on-chain in evm txn addition by hash
- Add and refactor tests to handle these changes.

Also, it fixes [`rotkehlchen/tests/conftest.py`](https://github.com/rotki/rotki/blob/2de468cae60907d1cfa685c4c98d9a0c5d891610/rotkehlchen/tests/conftest.py#L258) issue with the `caching_dir` filepath on Mac

### CODE REVIEW

File: rotkehlchen/chain/ethereum/node_inquirer.py [hunk 1]
```diff
@@ -281,17 +287,27 @@ def query_highest_block(self) -> BlockNumber:
...
             block_number=87042,
+            web3=web3,
         )
-        self.archive_connection = balance is not None and balance == FVal('5.1063307')
-        self.queried_archive_connection = True
-        return self.archive_connection
+        is_archive = balance == FVal('5.1063307')
+
+        try:
+            tx = web3.eth.get_transaction(HexStr('0x5c504ed432cb51138bcf09aa5e8a410dd4a1e204ef84bfed1be16dfba1b22060'))  # noqa: E501
```
Comment (nebolax) [inline rotkehlchen/chain/ethereum/node_inquirer.py:299]: I would move this tx hash and  the address used above to constants

File: rotkehlchen/chain/ethereum/node_inquirer.py [hunk 2]
```diff
@@ -281,17 +287,27 @@ def query_highest_block(self) -> BlockNumber:
...
+            block_no = tx['blockNumber']
+        except (
+            requests.exceptions.RequestException,
+            TransactionNotFound,
+            BlockchainQueryError,
+            KeyError,
+        ):
+            block_no = None
+        is_pruned = block_no != 46147
```
Comment (nebolax) [inline rotkehlchen/chain/ethereum/node_inquirer.py:308]: Why is there a need in checking block number? Won't it be sufficient to just check that the transaction was found?

File: rotkehlchen/chain/optimism/node_inquirer.py [hunk 3]
```diff
@@ -57,8 +62,27 @@ def query_highest_block(self) -> BlockNumber:
         log.debug('Optimism highest block result', block=block_number)
         return BlockNumber(block_number)
 
-    def have_archive(self, requery: bool = False) -> bool:
-        return False
+    def determine_capabilities(self, web3: Web3) -> tuple[bool, bool]:
```
Comment (nebolax) [inline rotkehlchen/chain/optimism/node_inquirer.py:65]: Looks like this code for determining capabilities could be abstracted.

File: rotkehlchen/externalapis/etherscan.py [hunk 4]
```diff
@@ -133,7 +144,7 @@ def _query(
             action: str,
             options: Optional[dict[str, Any]] = None,
             timeout: Optional[tuple[int, int]] = None,
-    ) -> Union[list[dict[str, Any]], str, list[EvmTransaction], dict[str, Any]]:
+    ) -> Union[list[dict[str, Any]], str, list[EvmTransaction], dict[str, Any], None]:
```
Comment (nebolax) [inline rotkehlchen/externalapis/etherscan.py:147]:
```suggestion
    ) -> Optional[Union[list[dict[str, Any]], str, list[EvmTransaction], dict[str, Any]]]:
```

File: rotkehlchen/externalapis/etherscan.py [hunk 5]
```diff
@@ -200,6 +211,9 @@ def _query(
             try:
                 result = json_ret.get('result', None)
                 if result is None:
+                    if action in ('eth_getTransactionByHash', 'eth_getTransactionReceipt'):
```
Comment (nebolax) [inline rotkehlchen/externalapis/etherscan.py:214]: Hmm why to silence the error in these cases?

File: rotkehlchen/chain/ethereum/modules/yearn/vaults.py [hunk 6]
```diff
@@ -556,18 +568,24 @@ def _get_vault_withdraw_events(
             )
             tx_hash = deserialize_evm_tx_hash(withdraw_event['transactionHash'])
             tx_receipt = self.ethereum.get_transaction_receipt(tx_hash)
+            if tx_receipt is None:
```
Comment (nebolax) [inline rotkehlchen/chain/ethereum/modules/yearn/vaults.py:571]: In `_query` you added a check so that for receipts and transactions pruned nodes are not queried. So is there a need to handle an empty receipt here?

Comment (prettyirrelevant) [reply rotkehlchen/chain/ethereum/modules/yearn/vaults.py:571]: An empty receipt is possible since a `tx_hash` might be valid but does not exist on-chain.

File: rotkehlchen/chain/ethereum/node_inquirer.py (hunk 1 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/ethereum/node_inquirer.py:299]: ah yes, thanks

File: rotkehlchen/chain/ethereum/node_inquirer.py (hunk 2 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/ethereum/node_inquirer.py:308]: I am not so familiar with the nitty-gritty of ethereum but I guess you're right

File: rotkehlchen/chain/evm/node_inquirer.py [hunk 7]
```diff
@@ -444,25 +456,33 @@ def _query(self, method: Callable, call_order: Sequence[WeightedNode], **kwargs:
...
-                TransactionNotFound,
                 BlockNotFound,
                 BadResponseFormat,
                 ValueError,  # Yabir saw this happen with mew node for unavailable method at node. Since it's generic we should replace if web3 implements https://github.com/ethereum/web3.py/issues/2448  # noqa: E501
             ) as e:
-                log.warning(f'Failed to query {node} for {str(method)} due to {str(e)}')
+                log.warning(f'Failed to query {node_info} for {str(method)} due to {str(e)}')
                 # Catch all possible errors here and just try next node call
                 continue
+            except TransactionNotFound:
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/evm/node_inquirer.py:484]: When a transaction is added by its hash, if the hash is valid but isn't present on-chain, the user sees a `RemoteError` saying.... **"we tried all nodes but couldn't make the query. Check logs"** which is the same error you'd get if all nodes are down.

...which isn't so helpful to the user. So yeah

File: rotkehlchen/externalapis/etherscan.py (hunk 5 above)
Comment (prettyirrelevant) [reply rotkehlchen/externalapis/etherscan.py:214]: Because if a transaction does not exist on-chain, Etherscan returns `{'result': null}` as a response

File: rotkehlchen/chain/evm/transactions.py [hunk 8]
```diff
@@ -567,11 +602,17 @@ def add_transaction_by_hash(
...
         - DeserializationError if there's an issue deserializing a value.
+        - InputError if the tx_hash or its receipt is not found on the blockchain.
         - pysqlcipher3.dbapi2.IntegrityError if the tx_hash is present in the db or address is not tracked.
         """  # noqa: E501
         dbevmtx = DBEvmTx(self.database)
         transaction = self.evm_inquirer.get_transaction_by_hash(tx_hash)
+        if transaction is None:
+            raise InputError('Transaction not found on chain.')
         receipt_data = self.evm_inquirer.get_transaction_receipt(tx_hash)
+        if receipt_data is None:
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/evm/transactions.py:613]: Since it's a user input... also checked its usage in the codebase.
Looks more appropriate

File: rotkehlchen/externalapis/etherscan.py (hunk 5 above)
Comment (nebolax) [reply rotkehlchen/externalapis/etherscan.py:214]: Yeah, but should then a Remote Error / Input Error be raised?

File: rotkehlchen/chain/ethereum/modules/yearn/vaults.py (hunk 6 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/yearn/vaults.py:571]: Yeah, I mean that this tx hash is retrieved from chain, so it is not something user input, and thus I would say we can rely on it's existence.

But anyway, I think having an extra check won't hurt.

File: rotkehlchen/chain/evm/node_inquirer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/evm/node_inquirer.py:484]: Mmm aha, yeah, makes sense :+1:

File: rotkehlchen/chain/evm/transactions.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/evm/transactions.py:613]: Yes, makes sense :+1:

One thing though -- receipt always exists if transaction exists, doesn't it? So I would say that there is no need in checking for receipt being None.

But is not something serious.

File: rotkehlchen/externalapis/etherscan.py (hunk 5 above)
Comment (prettyirrelevant) [reply rotkehlchen/externalapis/etherscan.py:214]: `RemoteError` would be caught `self._query()` of `EVMNodeInquirer` which is generic

File: rotkehlchen/tests/unit/test_ethereum_inquirer.py [hunk 9]
```diff
@@ -304,3 +307,88 @@ def test_get_blocknumber_by_time_subgraph(ethereum_inquirer):
 def test_get_blocknumber_by_time_etherscan(ethereum_inquirer):
     """Queries etherscan for known block times"""
     _test_get_blocknumber_by_time(ethereum_inquirer, True)
+
+
+@pytest.mark.parametrize(*ETHEREUM_NODES_PARAMETERS_WITH_PRUNED_AND_NOT_ARCHIVED)
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_ethereum_inquirer.py:312]: for this and the other tests in here if they do remote connections, they should use vcr believe.

File: rotkehlchen/tests/unit/test_evm_transactions.py [hunk 10]
```diff
@@ -1,4 +1,5 @@
 from unittest.mock import patch
+
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_evm_transactions.py:2]: why this space here? If not changing something don't touch it :smile:

File: rotkehlchen/tests/unit/test_optimism_inquirer.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
+import pytest
+
+from rotkehlchen.tests.utils.ethereum import wait_until_all_nodes_connected
+from rotkehlchen.tests.utils.optimism import OPTIMISM_NODES_PARAMETERS_WITH_PRUNED_AND_NOT_ARCHIVED
+
+
+@pytest.mark.parametrize(*OPTIMISM_NODES_PARAMETERS_WITH_PRUNED_AND_NOT_ARCHIVED)
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_optimism_inquirer.py:7]: same comment about vcr in this file

File: rotkehlchen/tests/utils/ethereum.py [hunk 12]
```diff
@@ -48,6 +48,17 @@
 ])
 ALCHEMY_TEST = 'https://eth-mainnet.alchemyapi.io/v2/ga1GtB7R26UgzjextaVpbaWZ49nSi2zt'
 
+PRUNED_AND_NOT_ARCHIVED_NODE = WeightedNode(
+    node_info=NodeName(
+        name='Public Node',
+        endpoint='https://ethereum.publicnode.com',
```
Comment (LefterisJP) [inline rotkehlchen/tests/utils/ethereum.py:54]: nice so you found a public pruned node?

**Review by LefterisJP — COMMENTED** [review]
Hey so approach looks okay but have some comments:

### Failed Tests

Related tests seem to fail.

`rotkehlchen/tests/unit/test_ethereum_inquirer.py::test_use_open_nodes`
`rotkehlchen/tests/unit/test_ethereum_inquirer.py::test_ethereum_nodes_prune_and_archive_status`

### Different version of the get functions

So as discussed in the PR we need a different version of those get_transaction and get_receipt functions. Either the old ones should be renamed to `maybe_get_xxx` and then create the `get_xxx` version which also internally does an assert and returns the non optional type. Or ... you can create a `get_and_check_xxx` function ... but hm ... I think using the `maybe_get_xxx` approach is cleaner.

### Tests

I see you added lots of tests that test the functionality. you should use vcr there to mock this, as it's really not something we should be querying remotes in the tests for. Either with vcr or a simple mock of the responses. But perhaps vcr is cleaner.

File: rotkehlchen/tests/utils/ethereum.py (hunk 12 above)
Comment (prettyirrelevant) [reply rotkehlchen/tests/utils/ethereum.py:54]: yeah, for Ethereum

File: rotkehlchen/chain/ethereum/node_inquirer.py [hunk 13]
```diff
@@ -281,17 +291,26 @@ def query_highest_block(self) -> BlockNumber:
 
         return BlockNumber(block_number)
 
-    def have_archive(self, requery: bool = False) -> bool:
-        if self.queried_archive_connection and requery is False:
-            return self.archive_connection
-
+    def _is_pruned(self, web3: Web3) -> bool:
+        try:
+            tx = web3.eth.get_transaction(HexStr(PRUNED_NODE_CHECK_TX_HASH.hex()))  # pylint: disable=no-member  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/node_inquirer.py:296]: this is not needed. The typing in web3.eth.get_transaction is wrong. What you are doing here is you are turning a hexbytes, to hexstr and then pass it in, but hexbytes is accepted.

Just pass it as it and add a type ignore and an explanation as here: https://github.com/rotki/rotki/blob/c6bded6e4e4b44aea7771f6de8a90d8b87c508c7/rotkehlchen/chain/evm/node_inquirer.py#L690

Should work.

File: rotkehlchen/chain/evm/node_inquirer.py [hunk 14]
```diff
@@ -668,26 +693,43 @@ def _get_transaction_receipt(
...
             tx_hash=tx_hash,
         )
 
+    def get_transaction_receipt(
+            self,
+            tx_hash: EVMTxHash,
+            call_order: Optional[Sequence[WeightedNode]] = None,
+    ) -> dict[str, Any]:
+        """Retrieves the transaction receipt for the tx_hash provided.
+        This method assumes the tx_hash is present on-chain.
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/node_inquirer.py:713]:
```suggestion
        This method assumes the tx_hash is present on-chain and we are connected to at least 1 node that can retrieve it.
```

File: rotkehlchen/chain/evm/node_inquirer.py [hunk 15]
```diff
@@ -703,17 +745,32 @@ def _get_transaction_by_hash(
...
             tx_hash=tx_hash,
         )
 
+    def get_transaction_by_hash(
+            self,
+            tx_hash: EVMTxHash,
+            call_order: Optional[Sequence[WeightedNode]] = None,
+    ) -> EvmTransaction:
+        """Retrieves information about a transaction from its hash.
+        This method assumes the tx_hash is present on-chain.
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/node_inquirer.py:765]:
```suggestion
        This method assumes the tx_hash is present on-chain and we are connected to at least one node that can retrieve it.
```

File: rotkehlchen/chain/evm/node_inquirer.py [hunk 16]
```diff
@@ -1103,6 +1160,18 @@ def _process_contract_info(
 
         return decoded_contract_info
 
+    def determine_capabilities(self, web3: Web3) -> tuple[bool, bool]:
+        """This method checks for the capabilities of a rpc node. This includes:
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/node_inquirer.py:1164]:
```suggestion
        """This method checks for the capabilities of an rpc node. This includes:
```

File: rotkehlchen/chain/optimism/node_inquirer.py [hunk 17]
```diff
@@ -57,8 +68,26 @@ def query_highest_block(self) -> BlockNumber:
         log.debug('Optimism highest block result', block=block_number)
         return BlockNumber(block_number)
 
-    def have_archive(self, requery: bool = False) -> bool:
-        return False
+    def _is_pruned(self, web3: Web3) -> bool:
+        try:
+            tx = web3.eth.get_transaction(HexStr(PRUNED_NODE_CHECK_TX_HASH.hex()))  # pylint: disable=no-member  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/optimism/node_inquirer.py:73]: same comment on the typing here as in ethereum/node_inquirer.py

File: rotkehlchen/tests/conftest.py [hunk 18]
```diff
@@ -235,7 +235,7 @@ def vcr_cassette_dir(request: pytest.FixtureRequest) -> str:
 
         # verify that the cassettes are present locally and up to date
         if (tests_caching_dir / '.git').exists() is False:
-            cmd = f'git clone https://github.com/rotki/test-caching.git {tests_caching_dir}'
+            cmd = f'git clone https://github.com/rotki/test-caching.git "{tests_caching_dir}"'
```
Comment (LefterisJP) [inline rotkehlchen/tests/conftest.py:238]: Can you move this change in a different commit and make it to bugfixes? We need it in oth places if it fixes something.

**Review by LefterisJP — COMMENTED** [review]
Some more comments and 1 related test fail: `rotkehlchen/tests/api/test_evm_transactions.py::test_evm_transaction_hash_addition`

**Review by LefterisJP — APPROVED** [review]
LGTM. The failed test is due to:

```
18/02/2023 15:28:44 UTC] WARNING rotkehlchen.chain.evm.node_inquirer Greenlet-1: Failed to query NodeName(name='etherscan', endpoint='', owned=False, blockchain=<SupportedBlockchain.ETHEREUM: 'ETH'>) for <bound method EvmNodeInquirer._get_transaction_receipt of <rotkehlchen.chain.ethereum.node_inquirer.EthereumInquirer object at 0x7f208b541b80>> due to ethereum Etherscan API request https://api.etherscan.io/api?module=proxy&action=eth_getTransactionReceipt&txhash=0x692f9a6083e905bdeca4f0293f3473d7a287260547f8cbccc38c5cb01591fcda&apikey=8JT7WQBB2VQP5C3416Y8X3S8GBA3CVZKP4 failed with HTTP status code 503 and text The service is unavailable.
```

One thing I thought of is you probably need to re-add the "not found" error code in the add transaction by hash endpoint since it's now implemented.
